language-plutus-core/src/Language/PlutusCore/Rename/Internal.hs

```diff
@@ -160,15 +157,8 @@ renameTermM (LamAbs ann name ty body)  =
     withFreshenedName name $ \nameFr -> LamAbs ann nameFr <$> renameTypeM ty <*> renameTermM body
 renameTermM (TyAbs ann name kind body) =
     withFreshenedName name $ \nameFr -> TyAbs ann nameFr kind <$> renameTermM body
-renameTermM (IWrap ann pat arg term)   =
-    IWrap ann <$> renameTypeM pat <*> renameTypeM arg <*> renameTermM term
-renameTermM (Apply ann fun arg)        = Apply ann <$> renameTermM fun <*> renameTermM arg
-renameTermM (Unwrap ann term)          = Unwrap ann <$> renameTermM term
-renameTermM (Error ann ty)             = Error ann <$> renameTypeM ty
-renameTermM (TyInst ann term ty)       = TyInst ann <$> renameTermM term <*> renameTypeM ty
-renameTermM (Var ann name)             = Var ann <$> renameNameM name
-renameTermM con@Constant{}             = pure con
-renameTermM bi@Builtin{}               = pure bi
+renameTermM x                          =
+    (traverseOf termSubterms renameTermM >=> traverseOf termSubtypes renameTypeM >=> traverseOf termVars renameNameM) x
```

Comment:
Does this do three passes?

Reply:
Depends what you mean by pass. It looks at all the one-step subterms, and calls itself recursively, it looks at at ll the one-step subtypes, and calls the type version, it looks at all the one-step names and renames them.

It might do case analysis on the term multiple times, I guess.

Reply:
I'm not sure if it's good to write it like that or not. The previous version is already pretty safe, but it also looks more efficient (and according to Vanessa's investigations it *is* more efficient). I'm on the fence.

Anyway, I think I would just avoid the compulsory `traverseOf` and write this as

```haskell
renameTermM x = termSubterms renameTerm x >>= termSubtypes renameTypeM >>= termVars renameM
```